src/library/scala/collection/immutable/SortedMap.scala

```diff
@@ -112,6 +112,8 @@ trait SortedMapOps[K, +V, +CC[X, +Y] <: Map[X, Y] with SortedMapOps[X, Y, CC, _]
     (previousValue, nextValue) match {
       case (None, None) => coll
       case (Some(_), None) => this.removed(key).coll
+      case (Some(v0: AnyRef), Some(v1: AnyRef)) if v0 eq v1 => coll // V may override equals, thus we can't use == on AnyRef
```

Comment:
I do not understand this part. Why would `==` fail with overriden `equals`? The overriden `equals` would be called even when accessing through `AnyRef`, would not it?

Reply:
If someone has a custom type, which answer `==` based on the custom equals, if we use `==` here it changes the behavior of that program.
I used `eq` to be sure this change is safe and doesn't change users code's behavior.

Reply:
I don't have a good example (as I would avoid overriding equals), but let's say you have a type `Color`.
it has 2 implementing classes. `RGB` & `RGBA`.
You override equals such that you can compare `RGB` and `RGBA` based on the visible colors alone.

Now you want to update a map, and replace an `RGB` with the equivalent `RGBA`.

If I don't check for `eq`, I actually don't replace that value.

Reply:
> if we use == here it changes the behavior of that program.

I am afraid I still do not understand. In what way would the behaviour be changed?

Reply:
That sounds like having a `case class AB(a: Double, b: Double)` and comparing only on `a`. Such design would seem broken to me - two values which have some observable differences, yet compare as equal. I am afraid you are right such cases can exist and would behave differently with the optimization you introduce. I am not sure how careful one has to be to maintain functionality for such strangely (I would even say wrongly) designed types.

Reply:
> > if we use == here it changes the behavior of that program.
> 
> I am afraid I still do not understand. In what way would the behaviour be changed?

I'll try to build upon my lousy `Color` example.
Given such `RGB` and `RGBA` values, we can think about something like:
```scala
type Pixel = (Int, Int)
def overrideWithAlpha(pixelToColor: Map[Pixel, Color], mask: List[(Pixel, Double)]): Map[Pixel, Color] = {
  mask.foldLeft(pixelToColor){ case (p2c, (pixel, alpha)) =>
    p2c.updateWith(pixel){
      case Some(c: Color) => Some(RGBA(c.red, c.green, c.blue, alpha))
      case none => none
    }
  }
}
```
in this case, since `c: Color` can be of type `RBG`, and thus only compate the visible red/gree/blue colors with RGBA,
even though this is a real update, if we just use `==`, it yields `true`, thus we need `eq` to get `false`:
```scala
scala> c == RGBA(c.red, c.green, c.blue, alpha)
val res0: Boolean = true

scala> c eq RGBA(c.red, c.green, c.blue, alpha)
val res1: Boolean = false
```

Reply:
> That sounds like having a `case class AB(a: Double, b: Double)` and comparing only on `a`. Such design would seem broken to me - two values which have some observable differences, yet compare as equal. I am afraid you are right such cases can exist and would behave differently with the optimization you introduce. I am not sure how careful one has to be to maintain functionality for such strangely (I would even say wrongly) designed types.

I agree.
But figured that being extra safe here cost almost nothing, and we gain the advantage of not breaking existing programs.
Users are usually aware that they shouldn't mess with `equals` on types that are used as keys, but here we operate on the value.

Reply:
Unfortunately, those considerations are irrelevant, because it is in fact impossible to distinguish `AnyRef`s from `AnyVal`s at run-time. All values that are `eq` will be caught by this clause, *and* all values that are `==` will be caught by the next one.

This clause can be kept, as it is in fact valid for `AnyVal`s as well. The second clause needs to be removed, however, since it would also capture `AnyRef`s.

Reply:
Also note that the rest of this file already uses the idiom `value1.asInstanceOf[AnyRef] eq value2.asInstanceOf[AnyRef]` for this kind of optimizations. We should use the same idiom here.

Reply:
You are correct, of course, sorry for being blind. We are talking about replacing **values** with equal **values**, not about equality on **keys**.  It is quite possible someone is using such classes as values in the map, and it is important to preserve current behaviour for them.

Reply:
@sjrd thanks, I wasn't aware it is impossible to distinguish at runtime.
Pushed a commit to change that.